pkg/services/ngalert/state/cache_test.go

```diff
@@ -32,3 +36,65 @@ func TestTemplateCaptureValueStringer(t *testing.T) {
 		})
 	}
 }
+
+func TestExpandTemplate(t *testing.T) {
+	cases := []struct {
+		name          string
+		text          string
+		alertInstance eval.Result
+		labels        data.Labels
+		expected      string
+		expectedError error
+	}{{
+		name:     "instance labels are expanded into $labels",
+		text:     "{{ $labels.instance }} is down",
+		labels:   data.Labels{"instance": "foo"},
+		expected: "foo is down",
+	}, {
+		name:          "missing instance label returns error",
+		text:          "{{ $labels.instance }} is down",
+		labels:        data.Labels{},
+		expectedError: errors.New("error executing template __alert_test: template: __alert_test:1:86: executing \"__alert_test\" at <$labels.instance>: map has no entry for key \"instance\""),
+	}, {
+		name: "values are expanded into $values",
+		text: "{{ $values.A.Labels.instance }} has value {{ $values.A }}",
+		alertInstance: eval.Result{
+			Values: map[string]eval.NumberValueCapture{
+				"A": {
+					Var:    "A",
+					Labels: data.Labels{"instance": "foo"},
+					Value:  ptr.Float64(10),
+				},
+			},
+		},
+		expected: "foo has value 10",
+	}, {
+		name: "missing label in $values returns error",
+		text: "{{ $values.A.Labels.instance }} has value {{ $values.A }}",
+		alertInstance: eval.Result{
+			Values: map[string]eval.NumberValueCapture{
+				"A": {
+					Var:    "A",
+					Labels: data.Labels{},
+					Value:  ptr.Float64(10),
+				},
+			},
+		},
+		expectedError: errors.New("error executing template __alert_test: template: __alert_test:1:86: executing \"__alert_test\" at <$values.A.Labels.instance>: map has no entry for key \"instance\""),
```

Comment:
I'm wondering if we can not error the whole template here, but just the block? Similar to https://github.com/grafana/grafana/pull/36823 https://github.com/grafana/grafana/pull/36823#issuecomment-883382152 , cc @davidmparrott

Reply:
(not a comment to block from merging to be clear, could change later I think).

Reply:
I am open to changing this but perhaps we should do it in a different issue? It will also be easier to refer back to in future as we have one issue where we talk about changing the `missingkey` option.